File: taskmanager/src/main/java/com/example/alexeyglushkov/taskmanager/task/SimpleTaskPool.java
package com.example.alexeyglushkov.taskmanager.task;

import android.os.Handler;
import android.os.Looper;
import android.util.Log;

import com.example.alexeyglushkov.tools.HandlerTools;

import junit.framework.Assert;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by alexeyglushkov on 30.12.14.
 */
public class SimpleTaskPool implements TaskPool {
    static final String TAG = "SimpleTaskPool";

    //TODO: think about weakref
    protected List<TaskPoolListener> listeners;
    private Handler handler;

    //TODO: think about a map
    private List<Task> tasks;

    private Object userData;

    public SimpleTaskPool(Handler handler) {
        tasks = new ArrayList<>();
        listeners = new ArrayList<>();
        setHandler(handler);
    }

    @Override
    public void setHandler(Handler handler) {
        if (this.handler != null) {
            checkHandlerThread();
        }

        this.handler = handler;
    }

    @Override
    public Handler getHandler() {
        return handler;
    }

    @Override
    public void addTask(final Task task) {
        if (!Tasks.isTaskReadyToStart(task)) {
            Log.d(TAG, "Can't put task " + task.getClass().toString() + " because it has been added " + task.getTaskStatus().toString());
            return;
        }

        // TaskPool must set Waiting status on the current thread
        task.getPrivate().setTaskStatus(Task.Status.Waiting);

        HandlerTools.runOnHandlerThread(handler, new Runnable() {
            @Override
            public void run() {
                Log.d(TAG, "addTaskOnThread");
                addTaskOnThread(task);
            }
        });
    }

    private void addTaskOnThread(Task task) {
        checkHandlerThread();

        task.addTaskStatusListener(this);
        tasks.add(task);

        triggerOnTaskAdded(task);
    }

    protected void triggerOnTaskAdded(Task task) {
        for (TaskPoolListener listener : listeners) {
            listener.onTaskAdded(this,task);
        }
    }

    public void onTaskStatusChanged(Task task, Task.Status oldStatus, Task.Status newStatus) {
        if (Tasks.isTaskCompleted(task)) {
            removeTask(task);
        }
    }

    @Override
    public void removeTask(final Task task) {
        HandlerTools.runOnHandlerThread(handler, new Runnable() {
            @Override
            public void run() {
                if (tasks.remove(task)) {
                    for (TaskPoolListener listener : listeners) {
                        listener.onTaskRemoved(SimpleTaskPool.this, task);
                    }
                }
            }
        });
    }

    @Override
    public Task getTask(String taskId) {
        checkHandlerThread();

        for (Task task : tasks) {
            if(task.getTaskId() != null && task.getTaskId().equals(taskId)) {
                return task;
            }
        }

        return null;
    }

    @Override
    public int getTaskCount() {
        checkHandlerThread();

        return tasks.size();
    }

    @Override
    public List<Task> getTasks() {
        checkHandlerThread();

        return tasks;
    }

    @Override
    public void setUserData(Object data) {
        this.userData = data;
    }

    @Override
    public Object getUserData() {
        return userData;
    }

    @Override
    public void addListener(TaskPoolListener listener) {
        listeners.add(listener);
    }

    @Override
    public void removeListener(TaskPoolListener listener) {
        listeners.remove(listener);
    }

    protected void checkHandlerThread() {
        Assert.assertEquals(Looper.myLooper(), handler.getLooper());
    }
}

File: taskmanager/src/main/java/com/example/alexeyglushkov/taskmanager/task/SimpleTaskPool.kt
package com.example.alexeyglushkov.taskmanager.task

import android.os.Handler
import android.os.Looper
import android.util.Log
import androidx.annotation.WorkerThread

import com.example.alexeyglushkov.tools.HandlerTools
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.launch

import org.junit.Assert

import java.util.ArrayList

/**
 * Created by alexeyglushkov on 30.12.14.
 */
// TODO: think about abstractTaskPool without tasks list to be able to inherit code in PriorityTaskProvider
open class SimpleTaskPool(scope: CoroutineScope) : TaskPool {

    //TODO: think about weakref
    protected var listeners = mutableListOf<TaskPool.Listener>() // TODO: set one listener

    private var _scope: CoroutineScope? = null
    override var scope: CoroutineScope
        get() = _scope!!
        set(value) {
            _scope = value
        }

    //TODO: think about a map
    protected val _tasks = mutableListOf<Task>()
    override var userData: Any? = null

    init {
        _scope = scope
    }

    //// Events

    override fun onTaskStatusChanged(task: Task, oldStatus: Task.Status, newStatus: Task.Status) {
        if (Tasks.isTaskCompleted(task)) {
            removeTask(task)
        }
    }

    /// Actions

    override fun addTask(task: Task) {
        if (task !is TaskBase) { assert(false); return }

        if (!Tasks.isTaskReadyToStart(task)) {
            Log.d(TAG, "Can't put task " + task.javaClass.toString() + " because it has been added " + task.taskStatus.toString())
            return
        }

        // TaskPool must set Waiting status on the current thread
        task.private.taskStatus = Task.Status.Waiting

        scope.launch {
            Log.d(TAG, "addTaskOnThread")
            addTaskOnThread(task)
        }
    }

    @WorkerThread
    private fun addTaskOnThread(task: Task) {
        if (task !is TaskBase) { assert(false); return }
        checkHandlerThread()

        task.addTaskStatusListener(this)

        var resultTask = task
        val taskId = task.taskId
        if (taskId != null) {
            val oldTask = getTask(taskId)
            val listener = listeners.firstOrNull()
            if (oldTask != null && listener != null) {
                resultTask = listener.onTaskConflict(this, task, oldTask)
            }
        }

        if (resultTask == task) {
            _tasks.add(task)
            triggerOnTaskAdded(task)
        }
    }

    @WorkerThread
    protected open fun triggerOnTaskAdded(task: Task) {
        for (listener in listeners) {
            listener.onTaskAdded(this, task)
        }
    }

    override fun removeTask(task: Task) {
        scope.launch {
            removeTaskOnThread(task)
        }
    }

    @WorkerThread
    private fun removeTaskOnThread(task: Task) {
        if (_tasks.remove(task)) {
            for (listener in listeners) {
                listener.onTaskRemoved(this@SimpleTaskPool, task)
            }
        }
    }

    override fun cancelTask(task: Task, info: Any?) {
        scope.launch {
            cancelTaskOnThread(task, info)
        }
    }

    @WorkerThread
    private fun cancelTaskOnThread(task: Task, info: Any?) {
        for (listener in listeners) {
            listener.onTaskCancelled(this@SimpleTaskPool, task, info)
        }
    }

    override fun addListener(listener: TaskPool.Listener) {
        listeners.add(listener)
    }

    override fun removeListener(listener: TaskPool.Listener) {
        listeners.remove(listener)
    }

    protected fun checkHandlerThread() {
        //Assert.assertEquals(Looper.myLooper(), this.handler.looper)
    }

    //// Getters

    @WorkerThread
    override fun getTaskCount(): Int {
        checkHandlerThread()
        return _tasks.size
    }

    @WorkerThread
    override fun getTasks(): List<Task> {
        checkHandlerThread()

        return _tasks
    }

    @WorkerThread
    override fun getTask(taskId: String): Task? {
        checkHandlerThread()

        for (task in _tasks) {
            if (task.taskId != null && task.taskId == taskId) {
                return task
            }
        }

        return null
    }

    companion object {
        internal val TAG = "SimpleTaskPool"
    }
}
